fix(gatsby-remark-images): don’t add links if image is already linked

jlengstorf opened the pull request:
Thanks for the help, @pieh!

fixes #6980


<!--
  Q. Which branch should I use for my pull request?
  A. Use `master` branch (probably).

  Q. Which branch if my change is an update to Gatsby v2?
  A. Definitely use `master` branch :)

  Q. Which branch if my change is an update to documentation or gatsbyjs.org?
  A. Use `master` branch. A Gatsby maintainer will copy your changes over to the `v1` branch for you

  Q. Which branch if my change is a bug fix for Gatsby v1?
  A. In this case, you should use the `v1` branch

  Q. Which branch if I'm still not sure?
  A. Use `master` branch. Ask in the PR if you're not sure and a Gatsby maintainer will be happy to help :)

  Note: We will only accept bug fixes for Gatsby v1. New features should be added to Gatsby v2.

  Learn more about contributing: https://www.gatsbyjs.org/docs/how-to-contribute/
-->

KyleAMathews commented:
Deploy preview for *using-postcss-sass* failed.

Built with commit af209f9215d8cd9c5ec2f3fda1cbdae1ec40027d

https://app.netlify.com/sites/using-postcss-sass/deploys/5b63a64073f2cf3a34e74d48

jlengstorf commented on packages/gatsby-remark-images/src/index.js:
```diff
@@ -30,11 +31,18 @@ module.exports = (
+    const inLink = ancestors.some(ancestor => ancestor.type === `link`)
+    if (!inLink) {
+      markdownImageNodes.push(node)
+    }
```
This is broken. We still need to fix the relative path, which I'm pretty sure @pieh told me to do and I just... didn't.

pieh replied:
I think we need to process images in both cases (when it's inside link and when it's not) - difference should be whether we wrap image in link here - https://github.com/jlengstorf/gatsby/blob/af209f9215d8cd9c5ec2f3fda1cbdae1ec40027d/packages/gatsby-remark-images/src/index.js#L175-L188

pieh commented:
Following trend of nonsense examples we had yesterday: ;)
Do we want to handle cases like:
```
<a href="https://example.org">![Mix #1](./images/gatsby.png)</a>
```
or
```
[<img alt="mix #2" src="./images/gatsby.png"/>](https://example.org)
```
with mix of markdown and html?

I would say this is not in scope for this PR - I think we just need to handle markdown image in markdown link here.

jlengstorf commented:
Alright — I think I cracked this. Added a check for both `link` and `<a>` in the ancestor tree, and added tests for goofy nesting and mixed markup/markdown. This fixes the problem in my example repo for combos like this:

```
[![Screenshot of video.](./images/gatsby.png)](https://example.org)

<a href="https://google.com">**![test](./images/gatsby.png)**</a>

[**<img src="./images/gatsby.png"/>**](https://gatsbyjs.org)
```
